service/history/historyEngine.go

```diff
@@ -1569,6 +1571,20 @@ func (e *historyEngineImpl) RespondActivityTaskFailed(
 				return nil, consts.ErrActivityTaskNotFound
 			}
 
+			cancelRequested = ai.CancelRequested
+
+			e.logger.Debug("RespondActivityTaskFailed", tag.WorkflowScheduleID(scheduleID), tag.ActivityInfo(ai), tag.NewBoolTag("cancelRequested", cancelRequested), tag.NewBoolTag("hasHeartBeatDetails", request.GetLastHeartbeatDetails() != nil))
+
+			if request.GetLastHeartbeatDetails() != nil && !request.GetLastHeartbeatDetails().Equal(ai.GetLastHeartbeatDetails()) {
```

Comment:
not sure if we should move this check into UpdateActivityProgress

Reply:
For the normal heartbeat api call the details payload could be the same or empty between calls just to update the `LastHeartbeatUpdateTime`. This could cause unintentional activity timeouts.

I my understanding correct? I was leaning toward removing this check after more testing.